src/ripple/app/misc/NetworkOPs.cpp

```diff
@@ -964,112 +1020,227 @@ Transaction::pointer NetworkOPsImp::processTransactionCb (
+}
+
+void NetworkOPsImp::doTransactionAsync (Transaction::pointer transaction,
+        bool bAdmin, FailHard failType)
+{
+    std::lock_guard<std::mutex> lock (mMutex);
 
-        getApp().getHashRouter ().setFlag (trans->getID (), SF_SIGGOOD);
+    mTransactions.push_back (TransactionStatus (transaction, bAdmin, false,
+        failType));
+
+    if (mDispatchState == DispatchState::none)
+    {
+        m_job_queue.addJob (jtBATCH, "transactionBatch",
+                std::bind (&NetworkOPsImp::transactionBatch, this));
+        mDispatchState = DispatchState::scheduled;
     }
+}
+
+void NetworkOPsImp::doTransactionSync (Transaction::pointer transaction,
+        bool bAdmin, FailHard failType)
+{
+    std::unique_lock<std::mutex> lock (mMutex);
 
+    mTransactions.push_back (TransactionStatus (transaction, bAdmin, true,
+        failType));
+
+    do
     {
-        auto lock = beast::make_lock(getApp().getMasterMutex());
+        if (mDispatchState == DispatchState::running)
+        {
+            // A batch processing job is already running, so wait.
+            mCond.wait (lock);
+        }
+        else
+        {
+            std::vector<TransactionStatus> transactions;
 
-        bool didApply;
-        TER r = m_ledgerMaster.doTransaction (
-            trans->getSTransaction (),
-            bAdmin ? (tapOPEN_LEDGER | tapNO_CHECK_SIGN | tapADMIN)
-            : (tapOPEN_LEDGER | tapNO_CHECK_SIGN), didApply);
-        trans->setResult (r);
+            mTransactions.swap (transactions);
+            mDispatchState = DispatchState::running;
+            lock.unlock();
+            apply (transactions);
+            lock.lock();
+            mDispatchState = DispatchState::none;
+            mCond.notify_all();
 
-        if (isTemMalformed (r)) // malformed, cache bad
-            getApp().getHashRouter ().setFlag (trans->getID (), SF_BAD);
+            if (mTransactions.size())
+            {
+                // More transactions need to be applied, but by another job.
+                m_job_queue.addJob (jtBATCH, "transactionBatch",
+                        std::bind (&NetworkOPsImp::transactionBatch, this));
+                mDispatchState = DispatchState::scheduled;
+            }
+        }
+    }
+    while (! transaction->getClearBatched());
+}
+
+void NetworkOPsImp::transactionBatch()
+{
+    std::unique_lock<std::mutex> lock (mMutex);
+
+    if (mDispatchState == DispatchState::running)
+        return;
+
+    mDispatchState = DispatchState::running;
+
+    while (mTransactions.size())
+    {
+        std::vector<TransactionStatus> transactions;
+
+        mTransactions.swap (transactions);
+        lock.unlock();
+        apply (transactions);
+        lock.lock();
+        mCond.notify_all();
+    }
+
+    mDispatchState = DispatchState::none;
+}
+
+void NetworkOPsImp::apply (std::vector<TransactionStatus>& transactions)
+{
+    Ledger::pointer ledger;
+    TransactionEngine engine;
```

Comment:
Pass in the lock, in an unlocked state. The called function creates the TransactionEngine on the stack, takes the lock, performs its activities, releases the lock, and then the TransactionEngine is destroyed